banzai/main.py

```diff
@@ -86,145 +90,154 @@ def get_stages_todo(last_stage=None, extra_stages=None):
+        images = stage_to_run.run(images)
 
+    output_files = image_utils.save_images(pipeline_context, images, master_calibration=calibration_maker)
+    return output_files
 
-def reduce_science_frames(pipeline_context):
-    stages_to_do = get_stages_todo()
-    reduce_frames_one_by_one(stages_to_do, pipeline_context)
 
+def process_directory(pipeline_context, raw_path, image_types=None, last_stage=None, extra_stages=None, log_message='',
+                      calibration_maker=False):
+    if len(log_message) > 0:
+        logger.info(log_message, extra_tags={'raw_path': raw_path})
+    stages_to_do = get_stages_todo(last_stage, extra_stages=extra_stages)
+    image_list = image_utils.make_image_list(raw_path)
+    image_list = image_utils.select_images(image_list, image_types,
+                                           pipeline_context.allowed_instrument_criteria,
+                                           db_address=pipeline_context.db_address)
+    if calibration_maker:
+        try:
+            run(stages_to_do, image_list, pipeline_context, calibration_maker=True)
+        except Exception as e:
+            logger.error(e, extra_tags={'raw_path': raw_path})
+    else:
+        for image in image_list:
+            try:
+                run(stages_to_do, [image], pipeline_context, calibration_maker=False)
+            except Exception as e:
+                logger.error(e, extra_tags={'filename': image})
 
-def reduce_experimental_frames(pipeline_context):
-    stages_to_do = get_stages_todo()
-    reduce_frames_one_by_one(stages_to_do, pipeline_context, image_types=['EXPERIMENTAL'])
 
+def parse_directory_args(pipeline_context, raw_path, extra_console_args=None):
+    if extra_console_args is None:
+        extra_console_args = []
 
-def reduce_experimental_frames_console():
-    run_end_of_night_from_console([reduce_experimental_frames], IMAGING_CRITERIA)
+    if pipeline_context is None:
+        if raw_path is None:
+            extra_console_args += [RAW_PATH_CONSOLE_ARGUMENT]
 
+        pipeline_context = parse_args(IMAGING_CRITERIA, extra_console_args=extra_console_args)
 
-def reduce_frames_one_by_one(stages_to_do, pipeline_context, image_types=None):
-    if image_types is None:
-        image_types = ['EXPOSE', 'STANDARD']
-    image_list = image_utils.make_image_list(pipeline_context)
-    original_filename = pipeline_context.filename
-    for image in image_list:
-        pipeline_context.filename = os.path.basename(image)
-        try:
-            run(stages_to_do, pipeline_context, image_types=image_types)
-        except Exception as e:
-            logger.error('{0}'.format(e), extra_tags={'filename': pipeline_context.filename,
-                                                          'filepath': pipeline_context.raw_path})
-    pipeline_context.filename = original_filename
+        if raw_path is None:
+            raw_path = pipeline_context.raw_path
+        return pipeline_context, raw_path
```

Comment:
I think this should be outside of the if statement (i.e. `parse_directory_args` will always return `pipeline_context` and `raw_path`)